### PR TITLE
Bb chapter 9

### PR DESCRIPTION
**Pull request by mb706**

### CODE REVIEW

**Comment by mb706** [comment]
not done yet, will go over one more section with bernd

**Comment by RaphaelS1** [comment]
I assume https://github.com/mlr-org/mlr3book/pull/688 and https://github.com/mlr-org/mlr3book/pull/620 are merged into this?

**Comment by mb706** [comment]
Yes. I can merge #688 again, but there are still some things to go over with bernd

File: book/chapters/chapter9/preprocessing.qmd [hunk 1]
```diff
@@ -12,36 +12,29 @@ In this book, preprocessing refers to everything that happens with *data* before
...
-Typically it can work well with natural language processing and computer vision problems, while standard tabular data lacks the structure for deep learning models to extract meaningful features automatically.
-Furthermore, different problems require different features to be extracted, and deep learning models may not always be able to identify the most relevant features for a given problem without human guidance.
-Hence, manual feature engineering is often required but with `mlr3pipelines`, we can simplify the process as much as possible.
+Typically it can work well with natural language processing and computer vision problems, while for standard tabular data, tree-based ensembles such as a random forest or gradient boosting are often still superior (and easier to handle). However, tabular deep learning approaches are currently catching up quickly.
+Hence, manual feature engineering is still often required but with `mlr3pipelines`, we can simplify the process as much as possible.
```
Comment (RaphaelS1) [inline book/chapters/chapter9/preprocessing.qmd:17]:
```suggestion
Hence, manual feature engineering is still often required but with `mlr3pipelines`, which can simplify the process as much as possible.
```

File: book/chapters/chapter9/preprocessing.qmd [hunk 2]
```diff
@@ -12,36 +12,29 @@ In this book, preprocessing refers to everything that happens with *data* before
 
 `r index('Data cleaning', aside = TRUE)`\index{exploratory data analysis|see{data cleaning}} is an important part of preprocessing that involves the removal of errors, noise, and redundancy in the data; we only consider data cleaning very briefly as it is usually performed outside of `mlr3` on the raw dataset.
 
-Another aspect of preprocessing is `r index('feature engineering', aside = TRUE)`, which covers all other transformations of data before it is fed to the machine learning model, including the creation of features from possibly unstructured data, such as written text, sequences or images.
-The goal of feature engineering is to prepare the data so that a model can be trained on it, and/or to further improve predictive performance.
-It is important to note that feature engineering helps mostly for simpler algorithms, while highly complex models usually gain less from it and require little data preparation to be trained.
-Common difficulties in data that can be solved with feature engineering include features with (high) skew distributions, high cardinality categorical features, missing observations, high dimensionality and imbalanced classes in classification tasks.
```
Comment (RaphaelS1) [inline book/chapters/chapter9/preprocessing.qmd:18]: @mb706 @berndbischl  you've removed all definition of feature engineering? Now this paragraph is hard to follow

File: book/chapters/chapter9/preprocessing.qmd [hunk 3]
```diff
@@ -111,25 +104,27 @@ names(which(lengths(tsk_ames$levels()) == 2))
...
 Compared to one-hot encoding, the main advantage is that only a single numeric feature is created regardless of the number of levels of the categorical features, hence it is especially useful for high-cardinality features.
 As information from the target is used to compute the impact scores, the encoding process must be embedded in cross-validation to avoid leakage between training and testing data (@sec-performance).
 
-As well as encoding features, other basic preprocessing steps for categorical features include removing constant features (which only have one level and may have been removed as part of data cleaning), and collapsing levels that occur very rarely as these may be missed during resampling, though stratification can be used to mitigate this (@sec-strat-group).
+As well as encoding features, other basic preprocessing steps for categorical features include removing constant features (which only have one level and may have been removed as part of data cleaning), and collapsing levels that occur very rarely.
+Unfortunately, these types of problems can occur as artifacts in resampling as dataset size is further reduced here.
```
Comment (RaphaelS1) [inline book/chapters/chapter9/preprocessing.qmd:126]:
```suggestion
These types of problems can occur as artifacts of resampling as the dataset size is further reduced.
```

File: book/chapters/chapter9/preprocessing.qmd [hunk 4]
```diff
@@ -161,12 +156,12 @@ bmr = benchmark(benchmark_grid(tsk_ames,
...
 ```
 
-In this small experiment, we see that the difference between the extended factor encoding pipeline and the simpler one-hot encoding strategy pipeline is only moderate.
+In this small experiment, we see that the difference between the extended factor encoding pipeline and the simpler one-hot encoding strategy pipeline is only very small.
 If you are interested in learning more about different encoding strategies, including a benchmark study comparing them, we recommend @pargent2022regularized.
 
 ## Missing Values {#sec-preprocessing-missing}
 
-A common problem in real-world data is `r index('missing data')`.
+A common problem in real-world data is `r index('missing values')` in features.
```
Comment (RaphaelS1) [inline book/chapters/chapter9/preprocessing.qmd:164]:
```suggestion
A common problem in real-world data is `r index('missing values', 'missing data')` in features.
```

File: book/chapters/chapter9/preprocessing.qmd [hunk 5]
```diff
@@ -244,26 +240,33 @@ bmr$aggregate(measure = msr_mae)[, .(learner_id, regr.mae)]
...
+2. `po("colapply")` -- Character and ordinal features are encoded as categorical, and date/time features are encoded as numeric.
+3. `po("imputehist")` -- Numeric features are imputed by histogram sampling.
+4. `po("imputesample")` -- Logical features are imputed by sampling from the empirical distribution -- this only affects the `$predict()`-step.
+5. `po("missind")` -- Missing data indicators are added for imputed numeric and logical variables.
+6. `po("imputeoor")` -- Missing values of categorical features are encoded with a new level.
+7. `po("fixfactors")` -- Fixes levels of categorical features such that the same levels are present during prediction and training (which may involve dropping empty factor levels).
+8. `po("imputesample")` -- Missing values in categorical features introduced from dropping levels in the previous step are imputed by sampling from the empirical distributions.
+9. `po("collapsefactors")` -- Categorical features levels are collapsed (starting from the rarest factors in the training data) until there are less than a certan number of levels.
+  This number is controlled by the `max_cardinality` argument and defaults to a conservative 1000.
```
Comment (RaphaelS1) [inline book/chapters/chapter9/preprocessing.qmd:263]:
```suggestion
9. `po("collapsefactors")` -- Categorical features levels are collapsed (starting from the rarest factors in the training data) until there are less than a certan number of levels, controlled by the `max_cardinality` argument (with a conservative default of `1000`).
```

File: book/chapters/chapter9/preprocessing.qmd [hunk 6]
```diff
@@ -244,26 +240,33 @@ bmr$aggregate(measure = msr_mae)[, .(learner_id, regr.mae)]
...
+10. `po("encode")` -- Categorical features are one-hot encoded.
+11. `po("removeconstants")` -- Constant features that might have been created in the previous steps are removed.
+
+`ppl("robustify")` has optional arguments `task` and `learner`.
+If these are provided, then the resulting pipeline will be set up to handle the given task and learner specifically: it will e.g. not impute missing values if the learner has the `"missings"` property, or if there are no missing values in the task to begin with.
```
Comment (RaphaelS1) [inline book/chapters/chapter9/preprocessing.qmd:268]:
```suggestion
If these are provided, then the resulting pipeline will be set up to handle the given task and learner specifically, for example, it will not impute missing values if the learner has the `"missings"` property, or if there are no missing values in the task to begin with.
```

File: book/chapters/chapter9/preprocessing.qmd [hunk 7]
```diff
@@ -432,13 +433,15 @@ tsk_ames_ext$data(1,
...
 learners = list(lrn_baseline, lrn("regr.rpart"), glrn_xgb_impact,
     glrn_rf_impute_oor, glrn_lm_robust, glrn_log_lm_robust)
 
 bmr_final = benchmark(benchmark_grid(c(tsk_ames_ext, tsk_ames), learners,
   rsmp_cv3))
-bmr_final$aggregate(measure = msr_mae)
+
+perf = bmr_final$aggregate(measure = msr_mae)
+perf[order(learner_id, task_id), .(task_id, learner_id, regr.mae)]
 ```
```
Comment (RaphaelS1) [inline book/chapters/chapter9/preprocessing.qmd:446]:
```suggestion
<!-- FIXME: RS READ -->
```

File: book/chapters/chapter9/preprocessing.qmd (hunk 2 above)
Comment (mb706) [reply book/chapters/chapter9/preprocessing.qmd:18]: I think this part got lost when transferring to gdoc and back, I will add it again

**Comment by mb706** [comment]
okay I am done with this one, too

File: book/chapters/chapter9/preprocessing.qmd [hunk 8]
```diff
@@ -13,9 +13,9 @@ In this book, preprocessing refers to everything that happens with *data* before
 `r index('Data cleaning', aside = TRUE)`\index{exploratory data analysis|see{data cleaning}} is an important part of preprocessing that involves the removal of errors, noise, and redundancy in the data; we only consider data cleaning very briefly as it is usually performed outside of `mlr3` on the raw dataset.
 
 Another aspect of preprocessing is `r index('feature engineering', aside = TRUE)`, which covers all other transformations of data before it is fed to the machine learning model, including the creation of features from possibly unstructured data, such as written text, sequences or images.
-The goal of feature engineering is to prepare the data so that a model can be trained on it, and/or to further improve predictive performance.
+The goal of feature engineering is to bring the data in a format that a given learner can handle, and/or to further improve predictive performance.
```
Comment (RaphaelS1) [inline book/chapters/chapter9/preprocessing.qmd:16]:
```suggestion
The goal of feature engineering is to enable the data to be handled by a given learner, and/or to further improve predictive performance.
```

**Review by RaphaelS1 — APPROVED** [review]
